File: server/read_savedata.js
#!/usr/bin/env node

var fs = require('fs');

function readRegionInfo(data, info, offset) {
  var region_info = {};
  var _offset = offset;
  region_info['region_id'] = data.readInt32LE(_offset += 4);
  region_info['active'] = Boolean(data.readInt32LE(_offset += 4));
  region_info['economy_bars'] = new Array();
  region_info['environ_bars'] = new Array();
  region_info['history'] = {};

  // _offset -= 4;
  // scores
  region_info['FUNDS']    = data.readFloatLE(_offset += 4);
  region_info['PC']       = data.readFloatLE(_offset += 4);
  region_info['EC']       = data.readFloatLE(_offset += 4);
  region_info['EN']       = data.readFloatLE(_offset += 4);
  region_info['CO2']      = data.readFloatLE(_offset += 4);
  region_info['AP']       = data.readFloatLE(_offset += 4);
  region_info['WP']       = data.readFloatLE(_offset += 4);
  region_info['LP']       = data.readFloatLE(_offset += 4);
  region_info['GDP']      = data.readFloatLE(_offset += 4);
  region_info['EQ']       = data.readFloatLE(_offset += 4);
  region_info['PP']       = data.readFloatLE(_offset += 4);
  region_info['TECH']     = data.readFloatLE(_offset += 4);
  region_info['GREEN']    = data.readFloatLE(_offset += 4);
  region_info['DONATION'] = data.readFloatLE(_offset += 4);
  
  // read economy bars
  for (var i = 0; i < info.calcCycles; i++) 
    region_info.economy_bars.push(data.readFloatLE(_offset += 4));

  // read economy bars
  for (var i = 0; i < info.calcCycles; i++) 
    region_info.environ_bars.push(data.readFloatLE(_offset += 4));

  // // read history 
  var num_nodes_in_history = data.readInt32LE(_offset += 4);
  // console.log (num_nodes_in_history);
  for (var i = 0; i < num_nodes_in_history; i++) {
    _offset += 1; // skip string length
    var node_key = data.toString('utf-8', _offset, _offset + 8); 
    var timestamp = data.readFloatLE(_offset += 8);
    _offset += 4;
    region_info.history[timestamp] = node_key;
  }
  info.regions.push(region_info);
  return _offset;
}

function readSaveData (req, res, filename, mappings, callback) {

  // console.log (filename);
  // create a read stream
  var reader = fs.createReadStream(filename);

  reader.on('data', function (data) {
    // console.log(data);
    var info = {};
    info['magic'] = data.readInt32LE(0);
    info['time'] = data.readFloatLE(4);
    info['expansionPnts'] = data.readFloatLE(8);
    info['calcCycles'] = data.readInt32LE(12);
    info['political_capital'] = data.readDoubleLE(16);
    info['funds'] = data.readDoubleLE(24);
    info['region_counts'] = data.readInt32LE(32);
    info['regions'] = new Array();

    var offset = 32;
    for (var i = 0; i < info.region_counts; i++) {
      offset = readRegionInfo(data, info, offset);
    }
    // console.log(info);

    if (callback != null) callback(filename);
    return res.render('savedata', {
      title : 'Environ',
      info : info,
      mappings : mappings
    });
  });
}

exports.read_savedata = readSaveData;
// readSaveData(null, null, "savedata.dat", null, null);

File: server/python/read_savedata.py
import sys
import json
from struct import unpack, calcsize

MAGIC = 1989127302

def read_content(fmt, fileContent):
    size = calcsize(fmt)
    return unpack(fmt, fileContent[:size])[0], fileContent[size:]

def read_contents(fmt, fileContent):
    size = calcsize(fmt)
    return unpack(fmt, fileContent[:size]), fileContent[size:]

def read_header(fileContent):
    """ format:
        i - Magic Number    int32
        f   - time
        d - Political Cap   double
        d - Funds           double
        i - numRegions      int32
    """
    # filetype check
    magic_number, fileContent = read_content("i", fileContent)
    assert magic_number == MAGIC, "Magic number corrupted"

    ret = dict()
    ret['time'], fileContent = read_content("f", fileContent)
    ret['political_capital'], fileContent = read_content("d", fileContent)
    ret['funds'], fileContent = read_content("d", fileContent)
    ret['region_counts'], fileContent = read_content("i", fileContent)

    ret['time'] = round(ret['time'], 2)
    ret['political_capital'] = round(ret['political_capital'], 2)
    ret['funds'] = round(ret['funds'], 2)
    # print json.dumps(ret)
    return ret, fileContent

def read_region(fileContent):
    region_id, fileContent = read_content("i", fileContent)
    region_active, fileContent = read_content("?", fileContent)

    # read scores
    scores  = ['FUNDS', 'PC', 'EC', 'EN', 'CO2', 'AP', 'WP', 'LP', 'GDP', 'EQ', 'PP', 'TECH', 'GREEN', 'DONATION']
    numbers, fileContent = read_contents("14f", fileContent)
    numbers = map(lambda x : round(x, 2), numbers)
    scores = dict(zip(scores, numbers))
    # print scores

    # read history nodes
    history = dict()
    num_nodes_in_history, fileContent = read_content("i", fileContent)
    assert num_nodes_in_history < 200, "too many nodes"
    for index in xrange(num_nodes_in_history):
        key, fileContent = read_content("10s", fileContent)
        activated_time, fileContent = read_content("f", fileContent)
        # print key.strip(), activated_time
        # history[key.strip()] = round(activated_time, 2)
        history[round(activated_time, 2)] = key.strip()

    ret = { 'id' : region_id, 'active' : region_active, 'scores' : scores, 'history' : history }
    # print json.dumps(ret)
    return ret, fileContent

if __name__ == '__main__':
    # info = read_xml()
    info = dict()
    regions = list()
    with open(sys.argv[1], mode='rb') as file: # b is important -> binary
    # with open("../data/savedata.dat", mode='rb') as file: # b is important -> binary
    # with open("../../testdata.dat", mode='rb') as file: # b is important -> binary
        fileContent = file.read()
        header, fileContent = read_header(fileContent)
        info.update(header)

        for index in xrange(info['region_counts']):
            region, fileContent = read_region(fileContent)
            regions.append(region)
    
    info['regions'] = regions
    print json.dumps(info, sort_keys=True)
